File: Problems/Leetcode/3001-4000/3001-3100/3098.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    static constexpr int mod = 1e9 + 7;

    int sumOfPowers(vector<int>& nums, int k) {
        int n = nums.size();
        ranges::sort(nums);

        int ans = 0;
        map<vector<int>, int> mp;
        function<int(int, int, int, int)> dfs = [&](int i, int cnt, int pre, int min_diff) -> int {
            if (mp.contains(vector<int> {cnt, pre, min_diff})) {
                return mp[vector<int> {cnt, pre, min_diff}];
            }

            if (cnt > i + 1) {
                return 0;
            }
            if (cnt == 0) {
                return min_diff;
            }

            int res1 = dfs(i - 1, cnt, pre, min_diff);
            int res2 = dfs(i - 1, cnt - 1, nums[i], min(min_diff, abs(pre - nums[i])));
            return mp[vector<int> {cnt, pre, min_diff}] = (res1 + res2) % mod;
        };

        return dfs(n - 1, k, INT_MAX / 2, INT_MAX / 2);
    }
};

File: Problems/Leetcode/3001-4000/3001-3100/3098.py
from collections import defaultdict
from functools import cache
from math import inf
from typing import List


class Solution:
    def sumOfPowers(self, nums: List[int], k: int) -> int:
        MOD = 1_000_000_007

        n = len(nums)
        nums.sort()

        # dp[i][p][v] 表示以 i 结尾时，长度为 p 且能量为 v 的子序列个数
        d = [[defaultdict(int) for _ in range(k + 1)] for _ in range(n)]
        res = 0
        for i in range(n):
            d[i][1][inf] = 1
            for j in range(i):
                diff = nums[i] - nums[j]
                for p in range(2, k + 1):
                    for v, cnt in d[j][p - 1]:
                        d[i][p][min(diff, v)] = (d[i][p][min(diff, v)] + cnt) % MOD
            for v, cnt in d[i][k]:
                res = (res + v * cnt % MOD) % MOD
        return res
